ast2vec/cloning.py

```diff
@@ -0,0 +1,178 @@
+import argparse
+import errno
+from functools import partial
+import json
+import logging
+from multiprocessing.dummy import Pool
+import os
+import shutil
+import subprocess as sb
+import zlib
+
+
+class LinguistFailedError(Exception):
+    """
+    Raised when we fail to classify the source files.
+    """
+    pass
+
+
+class RepoCloner:
+    def __init__(self, linguist, languages=None, log_level=logging.INFO, num_threads=1):
+        self._log = logging.getLogger("repo_cloner")
+        self._log.setLevel(log_level)
+        self._languages = languages
+        self._linguist, self._is_enry = self._find_enry(linguist)
+        self._num_threads = num_threads
+
+    def clone_repo(self, git_url, ignore, target_dir, redownload=False):
+        git_url = self._prepare_repo_url(git_url)
+
+        try:
+            repo_dir = self._prepare_repo_dir(git_url, target_dir, redownload)
+        except OSError as e:
+            if e.errno == errno.EEXIST:
+                self._log.info("%s already cloned, skipping.", git_url)
+                return
+            else:
+                raise
+
+        env = os.environ.copy()
+        env["GIT_TERMINAL_PROMPT"] = "0"
+        self._log.info("Cloning from %s...", git_url)
+
+        try:
+            sb.check_output(["git", "clone", "--depth=1", git_url, repo_dir],
+                            env=env, stdin=sb.DEVNULL, stderr=sb.STDOUT)
+        except sb.CalledProcessError as e:
+            shutil.rmtree(repo_dir, ignore_errors=True)
+            self._log.error("Git failed to clone repo. git stderr:\n\t" +
+                            "\n\t".join(e.output.decode("utf8").split("\n")))
+            if not ignore:
+                raise e from None
```

Comment:
Oh, Tim, Tim... All system calls like this can throw funny and completely unexpected exceptions. For example, when we debugged the zombie threads issue with @EgorBu, the number of processes in the system reached the limit and this call raised OSError with "resource exhausted". That happened because of the way Linux spawn a new process: fork() + exec() in the child. fork() fails before exec() and voila.